pycalphad/core/minimizer.pyx

```diff
@@ -600,6 +582,20 @@ cdef class SystemState:
                 for i in range(num_phase_dof+spec.num_statevars):
                     csst.moles_normalization_grad[i] += csst.mass_jac[comp_idx, i]
 
+    cdef double[::1] driving_forces(self, SystemSpecification spec):  # TODO: spec only used for # of components, is that something the State should know on its own?
```

Comment:
You can avoid passing in the spec by accessing `len(self.compsets[0].phase_record.nonvacant_elements)` if all you need is the number of components.

Reply:
You have to do `compset = self.compsets[0]` so that you get the fully-typed attribute access

Reply:
Or just `self.masses.shape[0]` which is initialized to be equal to `spec.num_components`

Reply:
This should be addressed now. I used the shape of the chemical potentials, since we use those to compute the driving forces anyway.